src/resource-manager.c

```diff
@@ -138,6 +138,65 @@ resource_manager_load_transient (ResourceManager  *resmgr,
     return rc;
 }
 /*
+ * This structure is used by the regap_session_callback to allow the
+ * response code from the callbacks to be passed to the caller.
+ */
+typedef struct {
+    ResourceManager *resmgr;
+    gboolean ret;
+} regap_session_data_t;
+/*
+ * This is a version of the 'regap_session' function with the with the
+ * appropriate prototype for use with GList 'foreach' iterator.
+ */
+void
+regap_session_callback (gpointer data_entry,
+                        gpointer data_user)
+{
+    regap_session_data_t *data = (regap_session_data_t*)data_user;
+    SessionEntry *entry  = SESSION_ENTRY (data_entry);
+
+    g_debug ("%s: SessionEntry 0x%" PRIxPTR, __func__, (uintptr_t)entry);
+    if (data->ret == TRUE) {
+        data->ret = regap_session (data->resmgr, entry);
+    } else {
+        g_critical ("%s: previous attempt to regap failed, skipping"
+                    " SessionEntry 0x%" PRIxPTR, __func__, (uintptr_t)entry);
+    }
+}
+/*
+ * This function is a handler for response codes that we may get from the
+ * TPM in response to commands. It may result in addtional commands being
+ * sent to the TPM so use it with caution.
+ */
+gboolean
+handle_rc (ResourceManager *resmgr,
+           TSS2_RC rc)
+{
+    regap_session_data_t data = {
+        .resmgr = resmgr,
+        .ret = TRUE,
+    };
+    gboolean ret;
+
+    g_debug ("%s: handling  RC 0x%" PRIx32, __func__, rc);
+    switch (rc) {
+    case TPM2_RC_CONTEXT_GAP:
+        g_debug ("%s: handling TPM2_RC_CONTEXT_GAP", __func__);
+        session_list_foreach (resmgr->session_list,
+                              regap_session_callback,
+                              &data);
+        ret = data.ret;
+        break;
+    default:
+        g_debug ("%s: Unable to recover gracefully from RC 0x%" PRIx32,
+                 __func__, rc);
+        ret = TRUE;
```

Comment:
If we can't gracefully recover, shouldn't we return `FALSE` here so that the calling site can propagate the error? Or are we returning `TRUE` and expecting the second `[load|save]_session()` to keep things sane? Sorry, I haven't had chance to dig into this deeply yet.

Reply:
By returning `TRUE` here we cause the RC to be sent back to the caller. Just because we don't know how to handle it doesn't mean they won't.

Reply:
Right, that makes sense. Thanks for clarifying!

Reply:
I spent some time looking back over this because my answer felt wrong. And it was. What will happen here if we fail to load the session:
- the session will be flushed
- the callers command that caused the RM to try to load the session will be sent to the TPM
- it will fail since it references a session that we couldn't load
- caller will get an RC from the TPM indicating that one of the sessions it references

If we want to handle everything in the RM we could construct a response object but sending the callers command through to the TPM and letting it create the RC seemed like the right thing to do at the time. Since we have all of the nice things provided by the type marshalling library (libtss2-mu) constructing response objects isn't nearly as much work as it used to be but even then packing out the right RC bits is a bit of work. I'm not against changing the approach here but I'd say it would make more sense in a future PR that makes the behavior consistent (we do similar things elsewhere).

Reply:
Thanks for looking back and this writing up. I'm OK with doing the work in a future PR, but could we file an enhancement request so that we don't lose this context?